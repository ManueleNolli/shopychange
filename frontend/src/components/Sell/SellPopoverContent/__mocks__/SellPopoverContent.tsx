import { divWithoutChildrenMock } from '../../../../../mocksUtils'

const SellPopoverContentMock = () =>
  divWithoutChildrenMock('SellPopoverContent')
export default SellPopoverContentMock
