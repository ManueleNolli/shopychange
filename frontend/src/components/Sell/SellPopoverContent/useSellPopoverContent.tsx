import { useEffect, useState } from 'react'

import {
  handlerQueryErrorCatch,
  isMarketplaceApprovedQuery,
} from '../../../services/BackendService'

import {
  askApproval,
  waitTransaction,
  createSale,
} from '../../../services/BlockchainService'

import { useUserContext } from '../../../context/userContext'

import { useToast } from '@chakra-ui/react'
import { SellPopoverProps } from '../../../types/props/SellPopoverProps'
import useLoading from '../../../hooks/useLoading'
import useFetching from '../../../hooks/useFetching'

export default function useSellPopoverContent({
  nft,
  onClose,
  onSucess,
}: SellPopoverProps) {
  const toast = useToast()
  const { userAddress, blockchainNetworkId } = useUserContext()
  const { isLoading, setIsLoading } = useLoading()
  const { isFetching, setIsFetching, isError, setIsError } = useFetching()
  const [price, setPrice] = useState<number>(0)
  const [isApproved, setIsApproved] = useState<boolean>(false)

  useEffect(() => {
    if (!userAddress || !blockchainNetworkId) return

    const fetchIsApproved = async () => {
      setIsFetching(true)
      const response = await handlerQueryErrorCatch(
        isMarketplaceApprovedQuery(
          blockchainNetworkId,
          nft.contractAddress,
          nft.tokenId
        )
      )
      if (response.error) {
        setIsError(true)
        setIsFetching(false)
        return
      }

      setIsApproved(response.result.isMarketplaceApproved)
      setIsFetching(false)
    }

    fetchIsApproved()
  }, [])

  const approve = async () => {
    if (!userAddress || !blockchainNetworkId) return

    setIsLoading(true)
    return (
      askApproval(blockchainNetworkId, nft.contractAddress, nft.tokenId)
        .then(waitTransaction)
        .then(() => {
          setIsLoading(false)
          setIsApproved(true)
        })
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        .catch((err: any) => {
          // First phrase of the error message is enough
          const errorMessage = err.message.split('.')[0]
          toast({
            variant: 'solid',
            title: 'Transaction failed',
            description: errorMessage,
            status: 'warning',
            duration: 9000,
            isClosable: true,
          })
          setIsLoading(false)
        })
    )
  }

  const sell = async () => {
    if (!userAddress || !blockchainNetworkId) return

    setIsLoading(true)
    return (
      createSale(blockchainNetworkId, nft.contractAddress, nft.tokenId, price)
        .then(waitTransaction)
        .then(() => {
          setIsLoading(false)
          onSucess()
          onClose()
        })
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        .catch((err: any) => {
          // First phrase of the error message is enough
          const errorMessage = err.message.split('.')[0]
          toast({
            variant: 'solid',
            title: 'Transaction failed',
            description: errorMessage,
            status: 'warning',
            duration: 9000,
            isClosable: true,
          })
          setIsLoading(false)
        })
    )
  }

  return {
    isFetching,
    isError,
    isApproved,
    isLoading,
    price,
    setPrice,
    approve,
    sell,
  }
}
