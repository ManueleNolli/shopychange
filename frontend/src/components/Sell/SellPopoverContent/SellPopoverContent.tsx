import React from 'react'

import {
  Flex,
  Center,
  ButtonGroup,
  Heading,
  Text,
  Spacer,
  Container,
  NumberInput,
  NumberInputField,
  ModalContent,
} from '@chakra-ui/react'

import LoadingButton from '../../LoadingButton/LoadingButton'
import ShadowButton from '../../ShadowButton/ShadowButton'

import { SellPopoverProps } from '../../../types/props/SellPopoverProps'

import { colors } from '../../../styles/Colors'
import PriceDistrubution from '../../PriceDistribition/PriceDistribution'
import useSellPopoverContent from './useSellPopoverContent'
import QueryContainer from '../../../utils/QueryContainer/QueryContainer'

export default function SellPopoverContent({
  nft,
  onClose,
  onSucess,
}: SellPopoverProps) {
  const {
    isApproved,
    isLoading,
    price,
    setPrice,
    approve,
    sell,
    isFetching,
    isError,
  } = useSellPopoverContent({
    nft,
    onClose,
    onSucess,
  })

  return (
    <ModalContent
      w={{ base: '80vw', md: '60vw' }}
      h={isApproved ? '60vh' : '30vh'}
    >
      <QueryContainer isLoading={isFetching} isError={isError}>
        <Flex direction="column" flex="1" p="1vh">
          <Spacer />

          <Container flex="1" centerContent>
            <Heading size="lg" textAlign={'center'}>
              Are you sure you want to sell <Text as="i">{nft.name}</Text>?
            </Heading>
          </Container>

          {isApproved ? (
            <>
              <Spacer />
              <Center>
                <NumberInput
                  defaultValue={0}
                  min={0}
                  step={0.01}
                  onChange={(value) => setPrice(parseFloat(value))}
                >
                  <NumberInputField
                    textAlign="center"
                    data-testid="price-input"
                  />
                </NumberInput>
                <Text pl="2" color={'gray.500'}>
                  eth
                </Text>
              </Center>
              <Spacer />
              <PriceDistrubution
                price={price}
                isOwner={true}
                address={nft.contractAddress}
                tokenId={nft.tokenId}
              />
            </>
          ) : null}

          <Spacer />

          <Center flex="1" mt="5">
            <ButtonGroup spacing="10">
              <LoadingButton isLoading={isLoading} data-testid="loading-button">
                {isApproved ? (
                  <ShadowButton
                    label={'Sell'}
                    bgColor={`${colors.sellChakraColor}.400`}
                    textColor={'white'}
                    shadowColor={colors.sellShadowColor}
                    hoverColor={`${colors.sellChakraColor}.500`}
                    focusColor={`${colors.sellChakraColor}.500`}
                    onClick={sell}
                  />
                ) : (
                  <ShadowButton
                    label={'Approve'}
                    bgColor={`${colors.approveChakraColor}.400`}
                    textColor={'white'}
                    shadowColor={colors.approveShadowColor}
                    hoverColor={`${colors.approveChakraColor}.500`}
                    focusColor={`${colors.approveChakraColor}.500`}
                    onClick={approve}
                  />
                )}
              </LoadingButton>
              <ShadowButton
                label={'Cancel'}
                bgColor={`${colors.cancelChakraColor}.500`}
                textColor={'white'}
                shadowColor={colors.cancelShadowColor}
                hoverColor={`${colors.cancelChakraColor}.600`}
                focusColor={`${colors.cancelChakraColor}.600`}
                onClick={onClose}
              />
            </ButtonGroup>
          </Center>
          <Spacer />
        </Flex>
      </QueryContainer>
    </ModalContent>
  )
}
