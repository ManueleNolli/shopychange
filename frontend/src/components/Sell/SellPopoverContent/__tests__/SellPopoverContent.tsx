import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import SellPopoverContent from '../SellPopoverContent'
import useSellPopoverContent from '../useSellPopoverContent'

jest.mock('@chakra-ui/react', () => {
  const originalChakraUI = jest.requireActual('@chakra-ui/react')
  return {
    ...originalChakraUI,
    ModalContent: ({ children }: { children: React.ReactNode }) => (
      <div data-testid="ModalContent">{children}</div>
    ),
  }
})

jest.mock('../useSellPopoverContent', () => jest.fn())
jest.mock('../../../PriceDistribition/PriceDistribution')

const mockProps = {
  nft: {
    contractAddress: '0x0000000000',
    tokenId: 0,
    image:
      'https://ipfs.io/ipfs/bafybeietntip4sq3xmuwiz3s43yb3klrpa3okgtg5deysmbd5hbizhrb2q',
    name: 'Test NFT',
  },
  onClose: jest.fn(),
}

const onUpdate = jest.fn()

describe('SellPopoverContent', () => {
  test('renders correctly isApproved = true', () => {
    ;(useSellPopoverContent as jest.Mock).mockReturnValueOnce({
      isApproved: true,
      isLoading: false,
      price: 0,
      setPrice: jest.fn(),
      approve: jest.fn(),
      sell: jest.fn(),
      isFetching: false,
      isError: false,
    })

    const tree = render(
      <SellPopoverContent
        nft={mockProps.nft}
        onClose={mockProps.onClose}
        onSucess={onUpdate}
      />
    )

    expect(tree.container).toMatchSnapshot()
  })

  test('renders correctly isApproved = false', () => {
    ;(useSellPopoverContent as jest.Mock).mockReturnValueOnce({
      isApproved: false,
      isLoading: false,
      price: 0,
      setPrice: jest.fn(),
      approve: jest.fn(),
      sell: jest.fn(),
      isFetching: false,
      isError: false,
    })

    const tree = render(
      <SellPopoverContent
        nft={mockProps.nft}
        onClose={mockProps.onClose}
        onSucess={onUpdate}
      />
    )

    expect(tree.container).toMatchSnapshot()
  })

  test('changing price calls setPrice', async () => {
    const setPrice = jest.fn()
    ;(useSellPopoverContent as jest.Mock).mockReturnValueOnce({
      isApproved: true,
      isLoading: false,
      price: 0,
      setPrice: setPrice,
      approve: jest.fn(),
      sell: jest.fn(),
      isFetching: false,
      isError: false,
    })

    const tree = render(
      <SellPopoverContent
        nft={mockProps.nft}
        onClose={mockProps.onClose}
        onSucess={onUpdate}
      />
    )

    const input = tree.getByTestId('price-input')

    fireEvent.change(input, { target: { value: '1' } })

    expect(setPrice).toBeCalledWith(1)
  })
})
