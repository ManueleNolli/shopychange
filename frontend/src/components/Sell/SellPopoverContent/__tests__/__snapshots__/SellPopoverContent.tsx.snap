// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SellPopoverContent renders correctly isApproved = false 1`] = `
<div>
  <div
    data-testid="ModalContent"
  >
    <div
      class="css-1wjy6p3"
    >
      <div
        class="css-17xejub"
      />
      <div
        class="chakra-container css-11uhf22"
      >
        <h2
          class="chakra-heading css-xi606m"
        >
          Are you sure you want to sell 
          <i
            class="chakra-text css-0"
          >
            Test NFT
          </i>
          ?
        </h2>
      </div>
      <div
        class="css-17xejub"
      />
      <div
        class="css-1ozwi6w"
      >
        <div
          class="chakra-button__group css-s2zvk0"
          data-orientation="horizontal"
          role="group"
        >
          <button
            class="chakra-button css-ha7ah3"
            data-testid="onClick-button"
            style="white-space: initial;"
            type="button"
          >
            Approve
          </button>
          <button
            class="chakra-button css-lnz0i6"
            data-testid="onClick-button"
            style="white-space: initial;"
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
      <div
        class="css-17xejub"
      />
    </div>
  </div>
</div>
`;

exports[`SellPopoverContent renders correctly isApproved = true 1`] = `
<div>
  <div
    data-testid="ModalContent"
  >
    <div
      class="css-1wjy6p3"
    >
      <div
        class="css-17xejub"
      />
      <div
        class="chakra-container css-11uhf22"
      >
        <h2
          class="chakra-heading css-xi606m"
        >
          Are you sure you want to sell 
          <i
            class="chakra-text css-0"
          >
            Test NFT
          </i>
          ?
        </h2>
      </div>
      <div
        class="css-17xejub"
      />
      <div
        class="css-gmuwbf"
      >
        <div
          class="chakra-numberinput css-e69dqy"
        >
          <input
            aria-valuemax="9007199254740991"
            aria-valuemin="0"
            aria-valuenow="0"
            aria-valuetext="0.00"
            autocomplete="off"
            autocorrect="off"
            class="chakra-numberinput__field css-fzwid6"
            data-testid="price-input"
            inputmode="decimal"
            pattern="[0-9]*(.[0-9]+)?"
            role="spinbutton"
            type="text"
            value="0.00"
          />
        </div>
        <p
          class="chakra-text css-td3r9l"
        >
          eth
        </p>
      </div>
      <div
        class="css-17xejub"
      />
      <div
        data-testid="PriceDistribution"
      />
      <div
        class="css-17xejub"
      />
      <div
        class="css-1ozwi6w"
      >
        <div
          class="chakra-button__group css-s2zvk0"
          data-orientation="horizontal"
          role="group"
        >
          <button
            class="chakra-button css-79ap6m"
            data-testid="onClick-button"
            style="white-space: initial;"
            type="button"
          >
            Sell
          </button>
          <button
            class="chakra-button css-lnz0i6"
            data-testid="onClick-button"
            style="white-space: initial;"
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
      <div
        class="css-17xejub"
      />
    </div>
  </div>
</div>
`;
